### PR TITLE
RecyclerView Adapter integration

### PR DESCRIPTION
**Pull request by FerrariBruno**
Finished the integration of the base recyclerview adapter. Used a multiple item recyclerview adapter. 

### Related Trello cards:
* https://trello.com/c/Qmnx7lnD

### Additional Info:
* Fixed some bugs on MovieFanFilterView and MovieFanImageView.

### UI preview:
<p align="center">
  
</p>

/cc @xmartlabs/android

### CODE REVIEW

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        public DetailedFilmViewHolder onCreateViewHolder(@NonNull ViewGroup parent) {
+          return new DetailedFilmViewHolder(inflateView(parent, R.layout.item_detailed_film));
+        }
+      };
+
+  @MainThread
+  void setItems(@NonNull List<Film> films) {
+    Stream.of(films)
+        .limit(DETAILED_FILM_VIEWHOLDER_LIMIT)
+        .forEach(film -> addItem(detailedItemType, film));
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: does `addItem` call `notifyDataSetChanged`?

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    Stream.of(films)
+        .skip(DETAILED_FILM_VIEWHOLDER_LIMIT)
+        .forEach(film -> addItem(collapsedItemType, film));
+
+    notifyDataSetChanged();
+  }
+
+  void clearData() {
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: isn't iin the base project? can you add it please?

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    @BindView(R.id.genres)
+    TextView genresTextView;
+    @BindView(R.id.popularity)
+    TextView popularityTextView;
+    @BindView(R.id.description)
+    TextView descriptionTextView;
+
+    DetailedFilmViewHolder(@NonNull View view) {
+      super(view);
+      ButterKnife.bind(this, view);
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: is it needed? doesn't the superclass call it?

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  static final class CollapsedFilmViewHolder extends SingleItemBaseViewHolder<Film> {
+    @BindView(R.id.round_poster)
+    MovieFanCircleImageView posterImageView;
+    @BindView(R.id.collapsed_title)
+    TextView titleTextView;
+
+    CollapsedFilmViewHolder(@NonNull View view) {
+      super(view);
+      ButterKnife.bind(this, view);
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: same question

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/MovieFanPageBaseFragment.java [hunk 5]
```diff
@@ -1,31 +1,91 @@
...
+
+  @Override
+  protected int getLayoutResId() {
+    return R.layout.fragment_movie_page;
+  }
+
+  @Override
+  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState){
+    super.onViewCreated(view, savedInstanceState);
+    initRecyclerView(view);
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/MovieFanPageBaseFragment.java]: the presenter has to call setupView or something like that

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import android.support.annotation.NonNull;
+import android.support.v7.widget.LinearLayoutManager;
+import android.support.v7.widget.RecyclerView;
+
+import lombok.Setter;
+
+/**
+ * Created by bruno on 1/5/18.
+ */
+public abstract class OnDemandRecyclerViewScrollListener extends RecyclerView.OnScrollListener {
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: the base project has this class

File: app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java [hunk 7]
```diff
@@ -86,7 +86,8 @@ public void onApplyButtonClicked() {
 
   @NonNull
   private Optional<Genre> getGenreFromSpinner() {
-    return ANY_GENRE.equals(genreSpinnerView.getSelectedItem()) ?
+    Object selectedItem = genreSpinnerView.getSelectedItem();
+    return selectedItem == null || ANY_GENRE.equals(selectedItem) ?
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java]: You could use
```java
return Optional.ofNullable(selectedItem)
.filter(ANY_GENRE::equals);
```

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 1 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: it calls `notifyItemInserted`

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 2 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: it is, it's called `clearAll`, I missed it :)

Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: it is in the base project, but not in the current version. Maybe it's due for a version bump?

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 3 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: yes it does, my bad

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 4 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: yup

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java (hunk 6 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: yup, great!

Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: same comment as before, it is in the base project, but not in the current version.

File: app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java (hunk 7 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java]: I think i can't, i have to return an `Optional<Genre>`

Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java]: yup, it's just missing the cast on `selectedItem`. It should've been `(Genre) selectedItem`

Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java]: doesn't work :$

Comment (matir91) [reply app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanFilterView.java]: Why? It should

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 2 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: i will point to the latest commit

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java (hunk 6 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: i will point to the latest commit

Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: it's not in bigbang's core, it's in app

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/MovieFanPageBaseFragment.java [hunk 8]
```diff
@@ -1,31 +1,95 @@
...
+  @Nullable
+  private OnDemandRecyclerViewScrollListener scrollListener;
+
+  @Override
+  protected int getLayoutResId() {
+    return R.layout.fragment_movie_page;
+  }
+
+  @Override
+  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState){
```
Comment (matir91) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/MovieFanPageBaseFragment.java]: delete it

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java (hunk 6 above)
Comment (matir91) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/OnDemandRecyclerViewScrollListener.java]: add a card to move it

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    Stream.of(films)
+        .skip(DETAILED_FILM_VIEWHOLDER_LIMIT)
+        .forEach(film -> addItemWithoutNotifying(collapsedItemType, film, true));
+
+    notifyDataSetChanged();
+  }
+
+  static final class DetailedFilmViewHolder extends SingleItemBaseViewHolder<Film> {
+    @NonNull
+    private static final String DELIMITER = ", ";
```
Comment (michaelEllisUy) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: not required imo

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    @BindView(R.id.title)
+    TextView titleTextView;
+    @BindView(R.id.genres)
+    TextView genresTextView;
+    @BindView(R.id.popularity)
+    TextView popularityTextView;
+    @BindView(R.id.description)
+    TextView descriptionTextView;
+
+    DetailedFilmViewHolder(@NonNull View view) {
```
Comment (michaelEllisUy) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: This is compulsory, right?

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  }
+
+  static final class CollapsedFilmViewHolder extends SingleItemBaseViewHolder<Film> {
+    @BindView(R.id.round_poster)
+    MovieFanCircleImageView posterImageView;
+    @BindView(R.id.collapsed_title)
+    TextView titleTextView;
+
+    CollapsedFilmViewHolder(@NonNull View view) {
```
Comment (michaelEllisUy) [inline app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: i guess this is compulsory as well?

File: app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanImageView.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import android.util.AttributeSet;
+
+import com.squareup.picasso.Picasso;
+import com.xmartlabs.moviefan.R;
+
+/**
+ * Created by bruno on 1/5/18.
+ */
+public class MovieFanImageView extends AppCompatImageView {
+  public MovieFanImageView(@NonNull Context context, @NonNull AttributeSet attrs) {
```
Comment (michaelEllisUy) [inline app/src/main/java/com/xmartlabs/moviefan/ui/views/MovieFanImageView.java]: sort according to the number of parameters received.

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 9 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: ok :)

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 10 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: yup

File: app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java (hunk 11 above)
Comment (FerrariBruno) [reply app/src/main/java/com/xmartlabs/moviefan/ui/main/FilmsRecyclerViewAdapter.java]: yup

**Pull request merged**